### PR TITLE
Speed up globbing relative to given directories

### PR DESCRIPTION
**Pull request by segiddins**
# Description:

This is picking up @tenderlove's improvements in https://github.com/rubygems/rubygems/pull/971.

Benchmark:

```ruby
class Hello
  def self.old_find_files_from_load_path glob # :nodoc:
    $LOAD_PATH.map { |load_path|
      Dir["#{File.expand_path glob, load_path}#{Gem.suffix_pattern}"]
    }.flatten.select { |file| File.file? file.untaint }
  end

  def self.aaron_find_files_from_load_path glob # :nodoc:
    search = "#{glob}#{Gem.suffix_pattern}"
    $LOAD_PATH.map { |load_path|
      Dir.chdir(load_path) {
        Dir[search].map! { |f| File.join load_path, f }
      }
    }.flatten.select { |file| File.file? file.untaint }
  end
  
  def self.new_find_files_from_load_path glob # :nodoc:
    glob_with_suffixes = "#{glob}#{Gem.suffix_pattern}"
    $LOAD_PATH.map { |load_path|
      Dir.glob(glob_with_suffixes, base: load_path).map! {|f| File.join(load_path, f) }
    }.flatten.select { |file| File.file? file.untaint }
  end
end

search = "minitest/*_plugin.rb"

Benchmark.ips do |x|
  x.report('old') do
    Hello.old_find_files_from_load_path search
  end

  x.report('aaron') do
    Hello.aaron_find_files_from_load_path search
  end
  
  x.report('new') do
    Hello.new_find_files_from_load_path search
  end
  
  x.compare!
end
```

Results (on Ruby 2.5):

```
Warming up --------------------------------------
                 old    94.000  i/100ms
               aaron   297.000  i/100ms
                 new   729.000  i/100ms
Calculating -------------------------------------
                 old    840.927  (±28.2%) i/s -      3.666k in   5.094465s
               aaron      2.843k (±19.2%) i/s -     13.365k in   5.012555s
                 new      7.457k (± 9.4%) i/s -     37.179k in   5.046804s

Comparison:
                 new:     7457.4 i/s
               aaron:     2842.6 i/s - 2.62x  slower
                 old:      840.9 i/s - 8.87x  slower
```

# Tasks:

- [ ] Describe the problem / feature
- [ ] Write tests
- [ ] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: lib/rubygems/util.rb [hunk 1]
```diff
@@ -116,4 +116,18 @@ def self.traverse_parents directory, &block
     end
   end
 
+  ##
+  # Globs for files matching +pattern+ inside of +directory+,
+  # returning absolute paths to the matching files.
+
+  if RUBY_VERSION >= "2.5"
```
Comment (hsbt) [inline lib/rubygems/util.rb]: It's better to inject this condition into `self.glob_files_in_dir` method.

Comment (segiddins) [reply lib/rubygems/util.rb]: But then we incur the cost of the condition every time the method is called

Comment (hsbt) [reply lib/rubygems/util.rb]: Can you show the benchmark about method call? I understand your opinion. But I didn't see that its cost is better than code style.

Comment (deivid-rodriguez) [reply lib/rubygems/util.rb]: I personally agree with @hsbt here.

Comment (MSP-Greg) [reply lib/rubygems/util.rb]: I suspect the cost of the condition will always be far less that the cost of reading a dir, SSD or not...

Comment (segiddins) [reply lib/rubygems/util.rb]: Conditional in the method seems to be about 85% as fast as pre-computing it

Comment (MSP-Greg) [reply lib/rubygems/util.rb]: Turning around, are conditionally defined methods really that bad?

**Comment by MSP-Greg** [comment]
Is it a fast system with a fast SSD, or an older system with a hard drive?  How many files searched, how many files match?  Etc...

**Comment by segiddins** [comment]
feel free to change it ¯\_(ツ)_/¯

**Comment by segiddins** [comment]
changes made

**Comment by segiddins** [comment]
@bundlerbot r+

**Pull request merged**